src/conn.ts

```diff
@@ -23,106 +28,364 @@ export class ConnOptions {
   //* filter for the main client as long as it is not overwritten by the client itself
   toServerBlackList: string[] = ['keep_alive'];
   //* filter for all attached clients if it is not overwritten by the client
-  toClientBlackList: string[] = ['keep_alive'];
+  toClientBlackList: string[] = ['keep_alive', 'unlock_recipes'];
+  //* Middleware to control packets being send to the client and server
+  toClientMiddleware?: PacketMiddleware[] = [() => { }];
+  toServerMiddleware?: PacketMiddleware[] = [() => { }];
+}
+
+export interface packetCanceler {
+  /** Has property .isCanceled: boolean indicating if the packet has been canceled by another middleware.
+   * Use `cancel(false)` to un-cancel the packet again.
+   */
+  (unCancel?: boolean): void
+  isCanceled: boolean
+}
+
+export interface packetUpdater {
+  (update?: boolean): void
+  isUpdated: boolean
+}
+
+/**
+ * Middleware manager for packets. Used to modify cancel or delay packets being send to the client or server.
+ */
+export interface PacketMiddleware {
+  /** Contains meta information about the packet that is triggered. See the properties for more info. */
+  (info: {
+    /** Direction the packet is going. Should always be the same direction depending on what middleware direction you
+     * are registering.
+     */
+    bound: 'server' | 'client',
+    /** Only 'packet' is implemented */
+    writeType: 'packet' | 'rawPacket' | 'channel',
+    /** Packet meta. Contains the packet name under `name` also see {@link PacketMeta} */
+    meta: PacketMeta
+  }, 
+  /** The client connected to this packet */ pclient: Client, 
+  /** Parsed packet data as returned by nmp */ data: any, 
+  /** A handle to cancel a packet from being send. Can also force a packet to be send */ cancel: packetCanceler, 
+  /** Indicate that a packet has been modified */ update: packetUpdater
+  ): void | Promise<void>;
 }
 
+const writeBuffer = false
+
 export class Conn {
   options: ConnOptions;
   bot: Bot & { recipes: number[] };
-  pclient: Client | undefined;
+  /** Contains the currently writing client or undefined if there is none */
+  writingPclient: Client | undefined;
+  /** Contains clients that are actively receiving packets from the proxy bot */
+  receivingPclients: Client[] = [];
+  /** Contains all connected clients. Even when they are not receiving or sending any packets */
   pclients: Client[] = [];
```

Comment:
why introduce two new fields, `receivingPClients` and `writingPClient`, when `pclients` and `pclient` respectively filled both of those roles?
`pclients` were all connected clients, which were automatically listening as i assume `receivingPClients` are now and
`pclient` was the single writing pclient

Reply:
I renamed pclient to writingPClient and pclients to receivingPClients to make it more readable.